embassy-imxrt/src/gpio.rs

```diff
@@ -0,0 +1,1056 @@
+        self.pin.set_function(func);
+    }
+
+    /// Is high?
+    #[must_use]
+    pub fn is_high(&self) -> bool {
+        !self.is_low()
+    }
+
+    /// Is low?
+    #[must_use]
+    pub fn is_low(&self) -> bool {
+        self.pin.block().b(self.pin.port()).b_(self.pin.pin()).read() == 0
+    }
+
+    /// Current level
+    #[must_use]
+    pub fn get_level(&self) -> Level {
+        self.is_high().into()
+    }
+
+    /// Wait until the pin is high. If it is already high, return immediately.
+    #[inline]
+    pub async fn wait_for_high(&mut self) {
+        InputFuture::new(self.pin.reborrow(), InterruptType::Level, Level::High).await;
+    }
+
+    /// Wait until the pin is low. If it is already low, return immediately.
+    #[inline]
+    pub async fn wait_for_low(&mut self) {
+        InputFuture::new(self.pin.reborrow(), InterruptType::Level, Level::Low).await;
+    }
+
+    /// Wait for the pin to undergo a transition from low to high.
+    #[inline]
+    pub async fn wait_for_rising_edge(&mut self) {
+        InputFuture::new(self.pin.reborrow(), InterruptType::Edge, Level::High).await;
+    }
+
+    /// Wait for the pin to undergo a transition from high to low.
+    #[inline]
+    pub async fn wait_for_falling_edge(&mut self) {
+        InputFuture::new(self.pin.reborrow(), InterruptType::Edge, Level::Low).await;
+    }
+
+    /// Wait for the pin to undergo any transition, i.e low to high OR high to low.
+    #[inline]
+    pub async fn wait_for_any_edge(&mut self) {
+        if self.is_high() {
+            InputFuture::new(self.pin.reborrow(), InterruptType::Edge, Level::Low).await;
+        } else {
+            InputFuture::new(self.pin.reborrow(), InterruptType::Edge, Level::High).await;
+        }
+    }
+
+    /// Return a new Flex pin instance with level sensing disabled.
+    ///
+    /// Consumes less power than a flex pin with sensing enabled.
+    #[must_use]
+    pub fn disable_sensing(self) -> Flex<'d, SenseDisabled> {
+        // Cloning the pin is ok since we consume self immediately
+        let new_pin = unsafe { self.pin.clone_unchecked() };
+        drop(self);
+        Flex::<SenseDisabled>::new(new_pin)
+    }
+}
+
+impl<'d> Flex<'d, SenseDisabled> {
+    /// New flex pin.
+    pub fn new(pin: Peri<'d, impl GpioPin>) -> Self {
+        pin.set_function(Function::F0)
+            .disable_analog_multiplex()
+            .disable_input_buffer();
+
+        Self {
+            pin: pin.into(),
+            _sense_mode: PhantomData::<SenseDisabled>,
+        }
+    }
+
+    /// Return a new Flex pin instance with level sensing enabled.
+    #[must_use]
+    pub fn enable_sensing(self) -> Flex<'d, SenseEnabled> {
+        // Cloning the pin is ok since we consume self immediately
+        let new_pin = unsafe { self.pin.clone_unchecked() };
+        drop(self);
+        Flex::<SenseEnabled>::new(new_pin)
+    }
+}
+
+/// Input pin
+pub struct Input<'d> {
+    pin: Flex<'d, SenseEnabled>,
+}
```

Comment:
Should this implement `Drop` to undo the `set_as_input`? In other hals to preserve that after drop you need to `mem::forget` it

Reply:
Same with Output as well, I guess I forgot to click save on that bit of review as well.

Reply:
I'm not sure what would be a good policy here. Keep the last settings or undo everything back to reset state? I guess I'll `mem::forget` it for now.

Reply:
Yes the pins would return to a floating state. stm32 for example: https://docs.embassy.dev/embassy-stm32/git/stm32u595vj/gpio/struct.Output.html

Reply:
Wait a minute, when `Input` dropped, the inner `Flex` will be dropped, which already has a `Drop` impl. This matches what stm32 is doing.

Reply:
Ah right I didn't see that.